docs: more usage examples

switowski opened the pull request:
- NEW Adds examples on how to perform identity checks and list missing
  mappings.

Signed-off-by: Sebastian Witowski witowski.sebastian@gmail.com

switowski committed:
cli: usage examples

* NEW Adds example of how to perform identity checks and list missing
  mappings.

* NEW Adds a test file with incorrect MARC21 mappings that should be
  skipped during the MARC21 identity check.

Signed-off-by: Sebastian Witowski <witowski.sebastian@gmail.com>

switowski commented:
@SamiHiltunen WDYT? Any other useful functions what we could add ?

switowski commented:
I see that there are already some examples in https://github.com/inveniosoftware/dojson/blob/master/dojson/cli/__init__.py, so I don't know if it's a common practice to add documentation in `__init__` or in docs ?
If in `__init__`, it would still be nice to add an example with `missing` command.

jirikuncar commented on docs/index.rst:
```diff
@@ -47,6 +47,23 @@ A simple example on how to convert MARCXML to JSON:
     [marc21.do(create_record(data)) for data in split_stream(open('/tmp/data.xml', 'r'))]
 
 
+The best way to check if you are not losing metadata during conversion is to perform identity check (from MARC21, through master format, to MARC21):
```
IMHO we should break the long lines in RST too.

switowski replied:
:+1:

switowski commented:
@jirikuncar , do you think it makes sense to write documentation now or is it a common knowledge to check the `__init__` files for examples and docs (fell free to close this PR, if it's too early to write documentation because the code will change and we risk that documentation will be outdated and confusing) ?

tiborsimko commented:
@switowski The usual technique would be to add a "Usage" section where several usage examples would be expanded.

jirikuncar commented on docs/index.rst:
```diff
@@ -47,6 +47,24 @@ A simple example on how to convert MARCXML to JSON:
+The best way to check if you are not losing metadata during conversion is to
+perform identity check (from MARC21, through master format, to MARC21):
+
+.. code:: shell
```
`s/shell/console/`

jirikuncar commented on docs/index.rst:
```diff
@@ -47,6 +47,24 @@ A simple example on how to convert MARCXML to JSON:
+
+.. code:: shell
+
+    dojson -l marcxml -d marcxml do marc21 do to_marc21 < input.xml > output.xml
```
please start with "`$`"

jirikuncar commented on docs/index.rst:
```diff
@@ -47,6 +47,24 @@ A simple example on how to convert MARCXML to JSON:
+
+.. code:: shell
+
+    dojson -i input.xml -l marcxml missing marc21
```
can you use any real file from tests?

switowski replied:
That's a good point, we actually can't since we have 100% coverage for the existing fields.
We can add a test file with fields that is not MARC21 compliant and see if they were removed.

jirikuncar commented:
@switowski the usage section can be added to `dojson/cli/__init__.py` and included in docs using `.. automodule:: dojson.cli`.

switowski commented:
I have moved the usage examples into `dojson/cli/__init__.py` as this code will be included in the documentation in the `Command line interface` section (https://github.com/inveniosoftware/dojson/blob/master/docs/index.rst)
I have added a new test file with incorrect MARC21 mappings that should be dropped during the identity check (we already have a dummy `999` field in `test_core.py` file that is used during tests, but I can't use this file in the cli example).
@jirikuncar let me know what you think.

jirikuncar commented on dojson/cli/__init__.py:
```diff
@@ -31,7 +35,16 @@
+
+    $ dojson -i tests/data/handcrafted/text_incorrect_marc21.xml -l marcxml \
+      missing marc21
+    # 049__, 2452_, 999__
```
IMHO the output can be without `#`

jirikuncar commented on dojson/cli/__init__.py:
```diff
@@ -31,7 +35,16 @@
 .. code-block:: console
 
     $ dojson -l marcxml -d marcxml do marc21 do to_marc21 < in.xml > out.xml
```
can we use any of our existing files in this example too?

switowski replied:
Sure. I used the `in.xml` as the input file doesn't really matter (on the contrary to the example with `missing` function, where it makes sense to point user to an input file with fields that are non-marc21 standard)

jirikuncar commented:
@switowski it looks good. Please have a look at my comments and rebase. Thanks

switowski commented:
@jirikuncar Updated

jirikuncar commented on dojson/cli/__init__.py:
```diff
@@ -30,8 +34,17 @@
-    $ dojson -l marcxml -d marcxml do marc21 do to_marc21 < in.xml > out.xml
-    $ diff in.xml out.xml
+    $ dojson -l marcxml -d marcxml do marc21 do to_marc21 < tests/data/test_1.xml > output.xml
+    $ xmllint --format out.xml | diff - in.xml
```
`s/out.xml/output.xml/`
`s/in.xml/tests/data/test_1.xml/`

switowski replied:
Ahh, sorry, stupid mistake :/
It's fixed now.

jirikuncar commented:
@switowski please rebase. LGTM.

tiborsimko commented on dojson/cli/__init__.py:
```diff
@@ -30,8 +34,17 @@
+    $ dojson -l marcxml -d marcxml do marc21 do to_marc21 < tests/data/test_1.xml > output.xml
+    $ xmllint --format output.xml | diff - tests/data/test_1.xml
+
+To get a list of fields that are missing mappings use the ``missing`` command:
```
The word "mappings" may be misleading here. People may think of Elasticsearch mappings.  Better say "rules" or "conversion rules".  Let's use the same vocabulary everywhere.

tiborsimko replied:
BTW, if we are amending docs, then we may want to do some more clean-up work:
- the TOC does not look nice due to section/subsection formatting
- the user talk vs developer talk is not strictly separated

I guess I'll make a separate issue for this.

jirikuncar replied:
missing mappings -> missing rule ?

tiborsimko commented on dojson/cli/__init__.py:
```diff
@@ -18,9 +18,13 @@
+Usage
+~~~~~
+
+For first example see ``schema``command that accept string argument containing
```
After "schema" a space is missing, which breaks output formatting.

jirikuncar commented:
Superseded by #117.
